### PR TITLE
2624 - Campaign Creation

### PR DESCRIPTION
**Pull request by onyekaa**
This is a re-opened PR and addresses custom campaign customization:
https:// jira.greenpeace.org/browse/PLANET-2624

- PLANET-2624 - Created Campaign Post Types
- Allows user to select Campaign Theme (Arctic, Antarctic etc)
- Created Campaign template files (both PHP and Twig files )
- Previous commits were squashed, including the removal of unneeded style files. Those were for the blocks and have been added to that repo.

### CODE REVIEW

**Comment by koyan** [comment]
Note: The Jira ticket for this is: https://jira.greenpeace.org/browse/PLANET-2624

**Comment by koyan** [comment]
@onyekaa please rebase your branch and resolve conflicts

**Comment by comzeradd** [comment]
Please squash these two commits as well. So the addition/removal of unneeded files is not part of the project's git log.

File: classes/class-p4-campaigns.php [hunk 1]
```diff
@@ -285,5 +293,138 @@ public function enqueue_admin_assets() {
...
+		 * @param $post
+		 */
+		public function campaign_page_templates_meta_box_callback( $post ) {
+
+			// Add a nonce field so we can check for it later.
+			wp_nonce_field( 'campaign_page_template_nonce', 'campaign_page_template_nonce' );
+
+			$value = get_post_meta( $post->ID, '_campaign_page_template', true );
+
+			$campaign_templates = array(
```
Comment (mikelittle) [inline classes/class-p4-campaigns.php]: I saw this same list in a different file. Should there be a generic function to supply these common values?

File: classes/class-p4-campaigns.php [hunk 2]
```diff
@@ -285,5 +293,138 @@ public function enqueue_admin_assets() {
...
+				'oceans'    => __( 'Oceans', 'planet4-master-theme-backend' ),
+				'oil'       => __( 'Oil', 'planet4-master-theme-backend' ),
+				'plastic'   => __( 'Plastics', 'planet4-master-theme-backend' ),
+			);
+
+	?>
+			<select id="campaign_page_template" name="campaign_page_template">
+				<option value=""><?php _e( 'Select Campaign Template', 'planet4-master-theme-backend' ); ?></option>
+				<?php foreach ( $campaign_templates as $campaign => $campaign_template ) {
+					$selected = $value == $campaign ? 'selected="selected"' : '';
```
Comment (mikelittle) [inline classes/class-p4-campaigns.php]: Why not use WordPress' standard `selected()` function?

File: classes/class-p4-campaigns.php [hunk 3]
```diff
@@ -285,5 +293,138 @@ public function enqueue_admin_assets() {
...
+
+		/**
+		 * callback function for campaign page template selection
+		 *
+		 * @param $post
+		 */
+		public function campaign_page_templates_meta_box_callback( $post ) {
+
+			// Add a nonce field so we can check for it later.
+			wp_nonce_field( 'campaign_page_template_nonce', 'campaign_page_template_nonce' );
```
Comment (mikelittle) [inline classes/class-p4-campaigns.php]: A nonce action should normally include something unique to this instance (e.g. post_id) to stop it being reused on another sibling item.

File: classes/class-p4-campaigns.php [hunk 4]
```diff
@@ -285,5 +293,138 @@ public function enqueue_admin_assets() {
...
+
+			/* OK, it's safe for us to save the data now. */
+
+			// Make sure that it is set.
+			if ( ! isset( $_POST['campaign_page_template'] ) ) {
+				return;
+			}
+
+			// Sanitize user input.
+			$campaign_page_template = sanitize_text_field( $_POST['campaign_page_template'] );
```
Comment (mikelittle) [inline classes/class-p4-campaigns.php]: `sanitize_text_field()` won't validate that the value is one of the acceptable (hard-coded) values. It may be better to use a common function (mentioned above) and validate with `in_array()`.

File: single-campaigns.php [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+// Get the cmb2 custom fields data
+// Articles block parameters to populate the articles block
+// p4_take_action_page parameter to populate the take action boxout block
+// Author override parameter. If this is set then the author profile section will not be displayed.
+$page_meta_data                 = get_post_meta( $post->ID );
+$campaign_template              = ! empty( $page_meta_data['_campaign_page_template'][0] ) ? $page_meta_data['_campaign_page_template'][0] : 'antarctic';
+$page_terms_data                = get_the_terms( $post, 'p4-page-type' );
+$background_image_id                = get_post_meta( get_the_ID(), 'background_image_id', 1 );
+$context['background_image']        = wp_get_attachment_url( $background_image_id );
+$context['background_image_srcset'] = wp_get_attachment_image_srcset( $background_image_id, 'full' );$take_action_page               = $page_meta_data['p4_take_action_page'][0] ?? '';
```
Comment (mikelittle) [inline single-campaigns.php]: Formatting is inconsistent and non-standard.

File: classes/class-p4-campaigns.php [hunk 6]
```diff
@@ -44,6 +49,10 @@ public function __construct() {
 		 * Class hooks.
 		 */
 		private function hooks() {
+
+			add_action( 'init', [ $this, 'register_campaigns_cpt' ] );
+			add_action( 'add_meta_boxes',  [ $this, 'campaign_page_templates_meta_box' ] );
+			add_action( 'save_post', [ $this, 'save_campaign_page_templates_meta_box_data' ] );
```
Comment (kirdia) [inline classes/class-p4-campaigns.php]: Although the hook contains many checks, I would suggest using a hook that will run only for the specific custom post type.
https://developer.wordpress.org/reference/hooks/save_post_post-post_type/

File: classes/class-p4-campaigns.php [hunk 7]
```diff
@@ -17,6 +17,11 @@ class P4_Campaigns {
 		 * @var string $taxonomy
 		 */
 		private $taxonomy = 'post_tag';
+
```
Comment (kirdia) [inline classes/class-p4-campaigns.php]: I would prefer all the related custom post type functions (hooks, metabox callbacks etc..) moved to another class.
The class name may be relevant (and confusing), but it is used currently for other purposes, so it is better to separate the cpt to it's own class (maybe P4_CampaignsPostType).

File: single-campaigns.php [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+// Build the shortcode for take action boxout block
+// Break the content to retrieve first 2 paragraphs and split the content if the take action page has been defined.
+if ( ! empty( $take_action_page ) ) {
+	$post->take_action_page   = $take_action_page;
+	$post->take_action_boxout = "[shortcake_take_action_boxout take_action_page='$take_action_page' /]";
+}
+
+// Build an arguments array to customize WordPress comment form.
+$comments_args = [
```
Comment (kirdia) [inline single-campaigns.php]: You can remove these lines, if the cpt won't support comments (although I see cpt supports comment).
https://github.com/greenpeace/planet4-master-theme/blob/3db348cdad85de31ce37b0e4da08987ae9b421b6/single-campaigns.php#L59-L82

Or use this in the single-campaigns twig template, if we are gonna use comments for this cpt
https://github.com/greenpeace/planet4-master-theme/blob/3db348cdad85de31ce37b0e4da08987ae9b421b6/templates/single.twig#L89

File: classes/class-p4-campaigns.php (hunk 7 above)
Comment (onyekaa) [reply classes/class-p4-campaigns.php]: Alright will work on this.

File: classes/class-p4-campaigns.php (hunk 1 above)
Comment (onyekaa) [reply classes/class-p4-campaigns.php]: Which file?

File: single-campaigns.php (hunk 8 above)
Comment (onyekaa) [reply single-campaigns.php]: Will do. Do we want comments on campaign pages?

File: classes/class-p4-campaigns.php (hunk 1 above)
Comment (mikelittle) [reply classes/class-p4-campaigns.php]: I saw what looks like the same list in classes/class-p4-campaigns.php around line 497 in PR#703
And the same common function could be used as part of the validation 
`if ( in_array( $post[something], array_keys( get_list_of_values() )  ) ) { `

File: gulpfile.js [hunk 9]
```diff
@@ -19,6 +19,7 @@ const path_scss = 'assets/scss/**/*.scss';
 const path_style = 'assets/scss/style.scss';
 const path_icons = 'images/icons/*.svg';
 const path_img = 'images/';
+
```
Comment (kirdia) [inline gulpfile.js]: Please remove this change.

File: single-campaigns.php (hunk 8 above)
Comment (kirdia) [reply single-campaigns.php]: No, I don't think so. You can remove this whole block for the moment.

File: single-campaigns.php [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+$context['filter_url'] = add_query_arg([
+	's'                                       => ' ',
+	'orderby'                                 => 'relevant',
+	'f[ptype][' . $context['page_type'] . ']' => $context['page_term_id'],
+	], get_home_url()
+);
+
+
+// Build the shortcode for articles block.
```
Comment (kirdia) [inline single-campaigns.php]: You can remove this block also.

File: single-campaigns.php [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+if ( 'yes' === $post->include_articles ) {
+	$post->articles = "[shortcake_articles exclude_post_id='" . $post->ID . "' /]";
+}
+
+// Build the shortcode for take action boxout block
+// Break the content to retrieve first 2 paragraphs and split the content if the take action page has been defined.
+if ( ! empty( $take_action_page ) ) {
```
Comment (kirdia) [inline single-campaigns.php]: You can remove this block also.

File: single-campaigns.php [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+$context['page_term_id']            = $page_terms_data[0]->term_id ?? '';
+$context['page_category']           = $category->name ?? __( 'Post page', 'planet4-master-theme' );
+$context['page_type_slug']          = $page_terms_data[0]->slug ?? '';
+$context['social_accounts']         = $post->get_social_accounts( $context['footer_social_menu'] );
+$context['og_title']                = $post->get_og_title();
+$context['og_description']          = $post->get_og_description();
+$context['og_image_data']           = $post->get_og_image();
+$context['custom_body_classes']     = 'brown-bg theme-' . $campaign_template;
+
+$context['filter_url'] = add_query_arg([
```
Comment (kirdia) [inline single-campaigns.php]: You can remove this block also.

File: single-campaigns.php [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+// Set Navigation Issues links.
+$post->set_issues_links();
+
+// Get the cmb2 custom fields data
+// Articles block parameters to populate the articles block
+// p4_take_action_page parameter to populate the take action boxout block
+// Author override parameter. If this is set then the author profile section will not be displayed.
+$page_meta_data                     = get_post_meta( $post->ID );
+$campaign_template                  = ! empty( $page_meta_data['_campaign_page_template'][0] ) ? $page_meta_data['_campaign_page_template'][0] : 'antarctic';
+$page_terms_data                    = get_the_terms( $post, 'p4-page-type' );
```
Comment (kirdia) [inline single-campaigns.php]: You can remove lines 20-21.

File: single-campaigns.php [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+$take_action_page                   = $page_meta_data['p4_take_action_page'][0] ?? '';
+$background_image_id                = get_post_meta( get_the_ID(), 'background_image_id', 1 );
+$context['background_image']        = wp_get_attachment_url( $background_image_id );
+$context['background_image_srcset'] = wp_get_attachment_image_srcset( $background_image_id, 'full' );
+$context['header_title']            = is_front_page() ? ( $page_meta_data['p4_title'][0] ?? '' ) : ( $page_meta_data['p4_title'][0] ?? $post->title );
+$context['header_subtitle']         = $page_meta_data['p4_subtitle'][0] ?? '';
+$context['header_description']      = wpautop( $page_meta_data['p4_description'][0] ?? '' );
+$context['header_button_title']     = $page_meta_data['p4_button_title'][0] ?? '';
+$context['header_button_link']      = $page_meta_data['p4_button_link'][0] ?? '';
+$context['page_type']               = $page_terms_data[0]->name ?? '';
```
Comment (kirdia) [inline single-campaigns.php]: You can remove lines 30, 31, 33.

File: classes/class-p4-post-campaign.php [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+
+		/** @var string $post_type */
+		private $post_type = 'campaigns';
+
+		/**
+		 * Page Types
+		 *
+		 * @var array $page_types
+		 */
+		public $page_types = [];
```
Comment (kirdia) [inline classes/class-p4-post-campaign.php]: You can remove this variable.

File: classes/class-p4-post-campaign.php [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+				'show_in_menu'       => true,
+				'query_var'          => true,
+				'rewrite'            => array( 'slug' => 'campaign' ),
+				'capability_type'    => 'post',
+				'has_archive'        => true,
+				'taxonomies'         => array( 'category', 'post_tag' ),
+				'hierarchical'       => false,
+				'menu_position'      => null,
+				'menu_icon'          => 'dashicons-megaphone',
+				'supports'           => array( 'title', 'editor', 'author', 'thumbnail', 'excerpt', 'comments' ),
```
Comment (kirdia) [inline classes/class-p4-post-campaign.php]: Let's remove the comments from here, for now.

**Comment by onyekaa** [comment]
@kirdia I don't understand. This branch is for the campaigns feature, with other things like customizations and export branching from this one. Or no?

**Comment by koyan** [comment]
@onyekaa what Kiriakos meant (and I just did) was that since is only part of the feature, and it cannot stand on its own, we should not merge it to develop, but to a dedicated branch. 
I have created a branch for the epic and changed the target of the PR towards that. 
Like that, when you fix the phpcs issues and the PR gets approved and merged, it will not polute the develop branch with a half-completed feature.

File: classes/class-p4-post-campaign.php [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+	 * Class P4_Post_Campaign
+	 */
+	class P4_Post_Campaign {
+
+		/**
+		 * Taxonomy
+		 *
+		 * @var string $taxonomy
+		 */
+		private $taxonomy = 'post_tag';
```
Comment (kirdia) [inline classes/class-p4-post-campaign.php]: @onyekaa please remove this variable, as it is not used.

File: classes/class-p4-post-campaign.php [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+
+			$args = array(
+				'labels'             => $labels,
+				'description'        => __( 'Campaigns', 'planet4-master-theme-backend' ),
+				'public'             => true,
+				'publicly_queryable' => true,
+				'show_ui'            => true,
+				'show_in_menu'       => true,
+				'query_var'          => true,
+				'rewrite'            => array( 'slug' => 'campaign' ),
```
Comment (kirdia) [inline classes/class-p4-post-campaign.php]: Please use short array notation here

File: classes/class-p4-post-campaign.php [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+				'capability_type'    => 'post',
+				'has_archive'        => true,
+				'taxonomies'         => array( 'category', 'post_tag' ),
```
Comment (kirdia) [inline classes/class-p4-post-campaign.php]: Please use short array notation here

File: classes/class-p4-post-campaign.php [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+				'hierarchical'       => false,
+				'menu_position'      => null,
+				'menu_icon'          => 'dashicons-megaphone',
+				'supports'           => array( 'title', 'editor', 'author', 'thumbnail', 'excerpt' ),
```
Comment (kirdia) [inline classes/class-p4-post-campaign.php]: and here also

**Comment by koyan** [comment]
@kirdia please do not merge until all the phpcs issues have also been fixed and the automatic checks pass

File: .stylelintrc [hunk 21]
```diff
@@ -64,7 +64,7 @@
     "selector-pseudo-class-case": "lower",
     "selector-pseudo-element-case": "lower",
     "selector-type-case": "lower",
-    "selector-type-no-unknown": true,
+    "selector-type-no-unknown": [true, { "ignoreTypes": ["/^_/","/^x/"] }],
```
Comment (comzeradd) [inline .stylelintrc]: This change seems unrelated to this PR, since it doesn't contain any css code changes.

**Comment by kirdia** [comment]
> @kirdia please do not merge until all the phpcs issues have also been fixed and the automatic checks pass


Yes, I missed that

@onyekaa there are code inspection errors also
You can check the errors locally using the repo's dependencies

1. Install repo dependencies
`composer install --dev`

2. Run phpcs on the console inside repo folder
`phpcs `

File: .stylelintrc (hunk 21 above)
Comment (onyekaa) [reply .stylelintrc]: I think Mehul added it back when he had the CSS as part of the commits, but I've removed it.

**Comment by comzeradd** [comment]
Can you also remove the two fonts and the svg icon from this PR, since you don't use them.

**Comment by kirdia** [comment]
> Can you also remove the two fonts and the svg icon from this PR? Since you don't use them.

@onyekaa can you please remove the extra files (fonts and and svg icon)? To move this PR forward.
Thank you!

**Comment by onyekaa** [comment]
@kirdia  and @koyan  done. Sorry for the delay. Here today to address issues.
